runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java

```diff
@@ -623,24 +764,191 @@ public IClientRouter getRouter(String address) {
+        int numThreads = parameters.nettyEventLoopThreads == 0
+                ? Runtime.getRuntime().availableProcessors() * 2 :
+                parameters.nettyEventLoopThreads;
+        ThreadFactory factory = new ThreadFactoryBuilder()
+                .setDaemon(true)
+                .setNameFormat(parameters.getPrefixedThreadFormat(
+                                    parameters.nettyEventLoopThreadFormat))
+                .setUncaughtExceptionHandler(this::handleUncaughtThread)
+                .build();
+        return parameters.socketType.getGenerator().generate(numThreads, factory);
+    }
+
+    /** Get a new {@link ExecutorService} for scheduling IO-bound tasks.
+     *
+     * @return  An {@link ExecutorService}.
+     */
+    private ExecutorService getNewIoExecutor() {
+        ThreadFactory factory = new ThreadFactoryBuilder()
+                .setDaemon(true)
+                .setNameFormat(parameters.getPrefixedThreadFormat(
+                        parameters.ioExecutorThreadFormat))
+                .setUncaughtExceptionHandler(this::handleUncaughtThread)
+                .build();
+        return Executors.newCachedThreadPool(factory);
+    }
+
+    /** Fetch a new layout, completing the layout in {@link this#layout}.
+     *
+     *  <p>The {@link this#layoutLock} ensures that there is only a single invocation of this
+     *  method at a time. When the method is executed, we fulfill the layout, or complete it
+     *  exceptionally if we could not retrieve the layout.
+     *
+     *  <p>This method tries to fetch a layout from each server in {@link this#layoutServers}
+     *  in random order. If all layout servers are unavailable, then we sleep for the duration
+     *  specified in {@link CorfuRuntimeParameters#connectionRetryRate} before retrying.
+     *
+     *  <p>The fetching thread can be stopped by setting {@link this#isShutdown} to true. This
+     *  will cause the fetching thread to terminate on the next retry with a
+     *  {@link UnrecoverableCorfuError}.
+     */
+    private void fetchLayout() {
+        if (layout.isDone()) {
+            log.warn("fetchLayout: Requested new layout but future already done!");
             return;
         }
-        layout = fetchLayout();
+
+        while (!isShutdown) {
+            // Every pass, we want to check using the current list of known
+            // layout servers in random order. We make sure to connect to
+            // all layout servers on the list in each pass.
+            List<String> serversToCheck = new ArrayList<>(layoutServers);
+            Collections.shuffle(serversToCheck);
+
+            for (String server : serversToCheck) {
+                try {
+                    // Get the layout from the server we chose.
+                    final Layout newLayout = getLayoutFromServer(server);
+
+                    // Update the epoch of all the active routers
+                    newLayout.getAllServers().forEach((name) -> {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove the parentheses around the "name" parameter [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1611)